### DISCUSSION TITLE
Procedure for using a third party icon library according to documentation not possible with TypeScript and Font Awesome

### DISCUSSION THREAD

**Discussion by manuelpoelzl** [chakra-ui/chakra-ui#5649] [Q&A]
### Description

When using a third party icon library (FontAwesome) in the "as" of the Chakra UI icon component, a type error occurs

### Link to Reproduction

https://codesandbox.io/s/modest-leaf-qlk5kd?file=/src/index.tsx

### Steps to reproduce

1. Install the React resources for FontAwesome
2. Import an icon from FontAwesome this way: import {faHouse} from "@fortawesome/free-solid-svg-icons";
3. Add a Chakra UI component to the React app
4. Set the "as" property to the previously imported icon

### Chakra UI Version

1.8.5

### Browser

Microsoft Edge 98.0.1108.56

### Operating System

- [ ] macOS
- [X] Windows
- [ ] Linux

**Comment by TimKolberger** [comment]
The linked codesandbox looks empty, could you kindly check?

**Comment by manuelpoelzl** [comment]
> The linked codesandbox looks empty, could you kindly check?

Here is a new link, it can be viewed there now.
https://codesandbox.io/s/silly-leavitt-y3v8cl?file=/src/index.tsx

**Comment by TimKolberger** [comment]
Kindly see the docs [how to use react-icons with Chakra UI](https://chakra-ui.com/docs/media-and-icons/icon#using-a-third-party-icon-library). I am not familiar with the package "@fortawesome/free-solid-svg-icons", but it looks like the imported `faHouse` is not a React component, which is a requirement to use the `as` prop.

**Reply by manuelpoelzl** [reply]
Thanks for your response!
I will try react-icons as described in the documentation. Maybe I can find a way to make it work with FontAwesome as well, but it shouldn't be necessary since FontAwesome is included in react-icons anyway.

Anyway, thanks for looking into this!

**Comment by DonaldLouch** [comment]
Probably not the best or optimal code but what I do for FontAwesome is:

```typescript

// config/fontAwesome

import { library } from '@fortawesome/fontawesome-svg-core'

import { fab } from '@fortawesome/free-brands-svg-icons'
import { far } from '@fortawesome/free-regular-svg-icons'
import { fas } from '@fortawesome/free-solid-svg-icons'

library.add(
    fab,
    far,
    fas
)

```

```typescript

// Component FIle

import { 
  Box,
  useColorModeValue
} from '@chakra-ui/react'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import '../config/fontAwesome'

...

export default function Component() {
  return (
    <Box
      color={useColorModeValue('LIGHTMODECOLOUR', 'DARKMODECOLOUR')} // To make it dark mode compatible/Change the colour of the icon
      fontSize="1.5rem" // Control the size of the icon
    >
      <FontAwesomeIcon icon={['{TYPE}', '{ICONNAME}']} width="100%" />
    </Box>
  )
}
```

**Reply by manuelpoelzl** [reply]
Thank you for your detailed example. I have tried it locally and it works fine.

However, I have now come to the conclusion that considering the effort involved, the react-icons package is currently the best option to use icons in Chakra UI, as it also supports the "as" property and no further steps are needed to get it running.

But it's good that we now have an example how it works without this package, should someone ask the same question in the future!

**Reply by DonaldLouch** [reply]
Very true 😊

**Comment by Pixelworlds** [comment]
In case anyone still needs this - @DonaldLouch has a great solution but it doesn't work when trying to use a Font Awesome icon with the `as` prop. For example, neither of the following work

```javascript
<ListIcon as={<FontAwesomeIcon icon={faCircleSmall} />} />
```
```javascript
<ListIcon as={faCircleSmall} />
```
However, you can use the `createIcon` method Chakra exposes to relatively easily convert the icon to something that `as` likes.

```javascript
const convertIcon = faIcon => {
  const { icon, iconName } = faIcon;

  return createIcon({
    displayName: iconName,
    viewBox: `0 0 ${icon[0]} ${icon[1]}`,
    d: icon[4],
  });
};
```
and then implement it 

```javascript
<ListIcon as={convertIcon(faCircleSmall)} />
```

**Comment by Jelle-vz** [comment]
What I did and what requires almost no configuration is the wrap `FontAwesomeIcon` in a chakra wrapper `chakra(FontAwesomeIcon)`. After that you can use it as you would normally.

```js
import { chakra } from '@chakra-ui/react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBars } from '@fortawesome/pro-solid-svg-icons';

export const myAwesomeComponent = () => {
  const Icon = chakra(FontAwesomeIcon);

  return (
      <Icon icon={faBars} marginRight={5} />
  )
}

```

**Comment by tecorraltd** [comment]
A couple helpers I find useful:

```
import { chakra } from "@chakra-ui/react";
import { IconDefinition } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const FaIcon = ({ icon }: { icon: IconDefinition }) => {
    const Icon = chakra(FontAwesomeIcon);

    return (
        <Icon icon={icon} />
    )
}

export default FaIcon;
```

```
import { IconButton, IconButtonProps } from "@chakra-ui/react"
import { IconDefinition } from "@fortawesome/fontawesome-svg-core"
import FaIcon from "./fa-icon"

type FaIconButtonProps = Omit<IconButtonProps, "icon"> & {
    icon: IconDefinition
}

const FaIconButton = (props: FaIconButtonProps) => {
    const {
        icon,
        ...otherProps
    } = props;

    return (
        <IconButton 
            icon={<FaIcon icon={icon} />}
            {...otherProps} 
        />
    )
}

export default FaIconButton;
```

**Comment by bhainesva** [comment]
Just adding a note based on some of the other answers. Using the `createIcon` helper function compared to using the normal `chakra` factory function will apply some different classes to the resulting output so the icon can render slightly differently depending on the approach. 

Some examples here: [CodeSandbox](https://codesandbox.io/p/sandbox/magical-monad-tw875z?file=%2Fsrc%2Findex.tsx%3A14%2C15&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt4nm75i00062e6b4r7x5w7o%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt4nm75i00022e6bl9x7f106%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt4nm75i00032e6b05bzkmdz%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt4nm75i00052e6b941hqt3c%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B44.850487123770684%252C55.149512876229316%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt4nm75i00022e6bl9x7f106%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt4nm75i00012e6bs7vd1qa9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A14%252C%2522startColumn%2522%253A15%252C%2522endLineNumber%2522%253A14%252C%2522endColumn%2522%253A15%257D%255D%257D%255D%252C%2522id%2522%253A%2522clt4nm75i00022e6bl9x7f106%2522%252C%2522activeTabId%2522%253A%2522clt4nm75i00012e6bs7vd1qa9%2522%257D%252C%2522clt4nm75i00052e6b941hqt3c%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt4nm75i00042e6bxsa297zb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clt4nm75i00052e6b941hqt3c%2522%252C%2522activeTabId%2522%253A%2522clt4nm75i00042e6bxsa297zb%2522%257D%252C%2522clt4nm75i00032e6b05bzkmdz%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clt4nm75i00032e6b05bzkmdz%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
